interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -63,63 +72,125 @@ public TrafficControlProxy() {
    *
    * @throws IOException
    */
-  public void setUp() throws IOException {
+  public void start() throws IOException {
     // ClientAcceptor uses a ServerSocket server so that the client can connect to the proxy as it
     // normally would a server. serverSock then connects the server using a regular Socket as a
     // client normally would.
     logger.info("Starting new proxy on port " + proxyPort);
     clientAcceptor = new ServerSocket();
+    clientAcceptor.setReuseAddress(true);
     clientAcceptor.bind(new InetSocketAddress(loopback, proxyPort));
     clientSock = clientAcceptor.accept();
     serverSock = new Socket();
     serverSock.connect(new InetSocketAddress(loopback, serverPort));
 
-    DataInputStream clientIn = new DataInputStream(clientSock.getInputStream());
-    DataOutputStream clientOut = new DataOutputStream(serverSock.getOutputStream());
-    DataInputStream serverIn = new DataInputStream(serverSock.getInputStream());
-    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
-
-    (new Thread(new ClientReader(clientIn))).start();
-    (new Thread(new ClientWriter(clientOut))).start();
-    (new Thread(new ServerReader(serverIn))).start();
-    (new Thread(new ServerWriter(serverOut))).start();
+    startWorkers();
   }
 
   public void shutDown() throws IOException {
     logger.info("Proxy shutting down... ");
     shutDown = true;
-    clientSock.close();
     clientAcceptor.close();
+    clientSock.close();
     serverSock.close();
+    // Have to also call interrupt because could be stuck in getServer/ClientChunk
+    for (Thread worker : workers) {
+      worker.interrupt();
+    }
+    logger.info("Shutdown Complete");
+  }
+
+  private void startWorkers() throws IOException {
+    DataInputStream clientIn = new DataInputStream(clientSock.getInputStream());
+    DataOutputStream clientOut = new DataOutputStream(serverSock.getOutputStream());
+    DataInputStream serverIn = new DataInputStream(serverSock.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    workers.add(new Thread(new ClientReader(clientIn)));
+    workers.add(new Thread(new ClientWriter(clientOut)));
+    workers.add(new Thread(new ServerReader(serverIn)));
+    workers.add(new Thread(new ServerWriter(serverOut)));
+
+    for (Thread worker : workers) {
+      worker.start();
+    }
+
   }
 
   private synchronized void incrementServerQueueSize(int delta) {
     serverQueueSize += delta;
+    if (serverQueueSize < queueLength) {
+      notifyAll();
+    }
   }
 
   private synchronized void incrementClientQueueSize(int delta) {
     clientQueueSize += delta;
+    if (clientQueueSize < queueLength) {
+      notifyAll();
+    }
+  }
+
+  private synchronized int getServerChunk() {
+    while (serverQueueSize == queueLength) {
+      try {
+        wait();
+      } catch (InterruptedException e) {
```

Comment:
This should actually propagate out of this method. By suppressing the interruption here, it seems like readers could block forever via a race with `shutdown` being set to `true`.